src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java

```diff
@@ -235,27 +226,105 @@ private void turnOffReadiness(Throwable exception){
-                LOGGER.warn("RabbitMQ blocked connection: {}", reason);
+        addBlockedListenersToConnection(this.connection);
+        addRecoveryListenerToConnection(this.connection);
+    }
+
+    private void addShutdownListenerToChannel(Channel channel, Boolean withRecovery) {
+        channel.addShutdownListener(cause -> {
+            LOGGER.debug("Closing the channel: ", cause);
+            if (!cause.isHardError() && cause.getReference() instanceof Channel) {
+                Channel channelCause = (Channel) cause.getReference();
+                Method reason = cause.getReason();
+                if (reason instanceof AMQImpl.Channel.Close) {
+                    var castedReason = (AMQImpl.Channel.Close) reason;
+                    if (castedReason.getReplyCode() != 200) {
+                        StringBuilder errorBuilder = new StringBuilder("RabbitMQ soft error occupied: ");
+                        castedReason.appendArgumentDebugStringTo(errorBuilder);
+                        errorBuilder.append(" on channel ");
+                        errorBuilder.append(channelCause);
+                        String errorString = errorBuilder.toString();
+                        LOGGER.warn(errorString);
+                        if (withRecovery && errorString.contains("PRECONDITION_FAILED")) {
+                            recoverSubscriptionsOfChannel(channel);
+                        }
+                    }
+                }
             }
+        });
+    }
 
-            @Override
-            public void handleUnblocked() throws IOException {
-                LOGGER.warn("RabbitMQ unblocked connection");
+    private void recoverSubscriptionsOfChannel(Channel channel) {
+        channelChecker.execute(() -> {
+            var mapEntries =
+                    channelsByPin
+                            .entrySet()
+                            .stream()
+                            .filter(entry -> Objects.nonNull(entry.getValue().channel) && channel.getChannelNumber() == entry.getValue().channel.getChannelNumber())
+                            .collect(Collectors.toList());
+            for (Map.Entry<PinId, ChannelHolder> mapEntry : mapEntries) {
+                PinId pinId = mapEntry.getKey();
+                SubscriptionCallbacks subscriptionCallbacks = mapEntry.getValue().subscriptionCallbacks;
+                if (subscriptionCallbacks != null) {
+
+                    try {
+                        LOGGER.info("Changing channel for holder with pin id: " + pinId.toString());
+
+                        channelsByPin.remove(pinId);
+                        channel.abort(400, "Aborted because of the recovery");
+                        basicConsume(pinId.queue, subscriptionCallbacks.deliverCallback, subscriptionCallbacks.cancelCallback);
+                    } catch (IOException | InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
```

Comment:
please remove the catch block, because this logic doesn't handle exception

Reply:
It is inside a callback. I added logging, but should we keep `throw new RuntimeException(e)`?